craft_application/util/repositories.py

```diff
@@ -0,0 +1,67 @@
+#  This file is part of craft-application.
+#
+#  Copyright 2023 Canonical Ltd.
+#
+#  This program is free software: you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License version 3, as
+#  published by the Free Software Foundation.
+#
+#  This program is distributed in the hope that it will be useful, but WITHOUT
+#  ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+#  SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public License
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Package repositories related functions."""
+
+from __future__ import annotations
+
+from pathlib import Path
+
+from craft_archives import repo  # type: ignore[import-untyped]
+from craft_archives.repo.package_repository import (  # type: ignore[import-untyped]
+    PackageRepositoryApt,
+    PackageRepositoryAptPPA,
+    PackageRepositoryAptUCA,
+)
+from craft_cli import emit
+from craft_parts import (
+    LifecycleManager,
+    ProjectInfo,
+)
+
+
+def install_package_repositories(
+    package_repositories: list[
+        PackageRepositoryApt | PackageRepositoryAptPPA | PackageRepositoryAptUCA
+    ]
+    | None,
+    lifecycle_manager: LifecycleManager,
+) -> None:
+    """Install package repositories in the environment."""
+    if not package_repositories:
+        emit.debug("No package repositories specified, none to install.")
+        return
+
+    repos = [repo.marshal() for repo in package_repositories]
```

Comment:
The fact that we have to marshal() the repositories here just so `repo.install()` re-unmarshals them makes me think this PR is trying to do too much now. In my opinion we should:

- Revert this bit and keep `package-repositories` as an optional list of dicts, as before;
- Release a new major version of craft-archives which changes the parameter type for `repo.install()` and associates;
- Do a new PR here to change the code to have this new package-repositories model.

What do you think @cmatsuoka and @syu-w ?

Reply:
Although I don't like the marshal and re-unmarshals parts, but this allows we check the project's configuration during early stage instead of a later runtime error in a VM or LXD. Unless we copy all the checks from craft-archives.

Reply:
So the actual reason is that the repositories API wants them in that format and it's convenient to maintain and validate separately. I agree with Tiago, let's keep the dictionary for now and see if there's a better solution later.